### PR TITLE
mgr/dashboard: fix rgw service creation in dashboard

### PR DESCRIPTION
**Pull request by aaSharma14**
In the rgw service creation form provide an option to add the realm, zonegroup and zone name in separate fields rather than entering a pattern in the service id

https://tracker.ceph.com/issues/61252

[Screencast from 2023-05-31 18-56-07.webm](https://github.com/ceph/ceph/assets/66050535/98319a89-f0bf-44af-b2ee-505ffd91ae03)

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by Pegonzal — COMMENTED** [review]
@aaSharma14 unit test failing over here

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html [hunk 1]
```diff
@@ -83,15 +83,54 @@
...
+          <label class="cd-col-form-label"
+                 for="zone_name">
+            <span i18n>Zone Name</span>
+          </label>
+          <div class="cd-col-form-input">
+            <input id="zone_name"
+                   class="form-control"
+                   type="text"
+                   formControlName="zone_name">
+          </div>
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:131]: Lets set the default value to "default" for zone and zonegroup, that's what is used.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts [hunk 2]
```diff
@@ -559,12 +553,16 @@ export class ServiceFormComponent extends CdForm implements OnInit {
...
-      const svcIdMatch = values['service_id'].match(this.RGW_SVC_ID_PATTERN);
-      svcId = svcIdMatch[1];
-      if (svcIdMatch[3]) {
-        serviceSpec['rgw_realm'] = svcIdMatch[3];
-        serviceSpec['rgw_zone'] = svcIdMatch[4];
+      if (
+        values['realm_name'] !== null &&
+        values['zone_name'] !== null &&
+        values['zonegroup_name'] !== null
+      ) {
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:560]: why do we have a `&&` condition, even one value is set we want to reflect that in service.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts [hunk 3]
```diff
@@ -559,12 +553,16 @@ export class ServiceFormComponent extends CdForm implements OnInit {
...
+        serviceSpec['rgw_realm'] = values['realm_name'];
+        serviceSpec['rgw_zonegroup'] = values['zonegroup_name'];
+        serviceSpec['rgw_zone'] = values['zone_name'];
       }
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:564]:
```suggestion
        serviceSpec['rgw_realm'] = values?.realm_name || null;
        serviceSpec['rgw_zonegroup'] = values?.zonegroup_name || null];
        serviceSpec['rgw_zone'] = values?.zone_name || null;
```

**Review by cloudbehl — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html (hunk 1 above)
Comment (cloudbehl) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:131]: Also add some helper text to each one of them

Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:131]: Done

**Review by Pegonzal — COMMENTED** [review]
I'm not a fan of the current implementation of the create realm, zonegroup and zone from the form

The text and the button are wierdly placed.
The text is redundant, `You can create a realm/zonegroup/zone by clicking here` next to the `Create Realm/Zonegroup/Zone` button.
The final text says: `By default the default realm/zonegroup/zone will be selected` which I also feel is redundant. If we are already filling the form fields with the default realm, zonegroup and zone is it really necessary the text?

A part from that nowhere is specified that the realm, zgp and zn we are creating will be set as the default ones.

Maybe something like:
image

Or directly giving the option to type in the field and create it if it doesnt exist would work better, like we have in other places such as creating an alert silence

For some reason my zonegroup and zones fields are not being filled by default
[screen-capture (10).webm](https://github.com/ceph/ceph/assets/106810838/3243c119-8a46-49c6-9901-9cea5cc874ed)

Also after creating the realm, zonegroup and zone
[screen-capture (11).webm](https://github.com/ceph/ceph/assets/106810838/97071f36-8562-4e0e-8378-c1844f511a6a)

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 4]
```diff
@@ -126,6 +127,184 @@ def migrate(self, realm_name=None, zonegroup_name=None, zone_name=None,
...
+                    zone_info = self.get_zone(rgw_zone)
+                    zones_info.append(zone_info)
+                    all_zones_info['zones'] = zones_info  # type: ignore
+            else:
+                all_zones_info['zones'] = []
+        if 'default_info' in rgw_zone_list and rgw_zone_list['default_info'] != '':
+            all_zones_info['default_zone'] = rgw_zone_list['default_info']  # type: ignore
+        else:
+            all_zones_info['default_zone'] = ''  # type: ignore
+        return all_zones_info
```
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/controllers/rgw.py]: Updated the PR

**Comment by aaSharma14** [comment]
> I'm not a fan of the current implementation of the create realm, zonegroup and zone from the form
> 
> The text and the button are wierdly placed. The text is redundant, `You can create a realm/zonegroup/zone by clicking here` next to the `Create Realm/Zonegroup/Zone` button. The final text says: `By default the default realm/zonegroup/zone will be selected` which I also feel is redundant. If we are already filling the form fields with the default realm, zonegroup and zone is it really necessary the text?
> 
> A part from that nowhere is specified that the realm, zgp and zn we are creating will be set as the default ones.
> 
> Maybe something like: image
> 
> Or directly giving the option to type in the field and create it if it doesnt exist would work better, like we have in other places such as creating an alert silence
> 
> For some reason my zonegroup and zones fields are not being filled by default [screen-capture (10).webm](https://github.com/ceph/ceph/assets/106810838/3243c119-8a46-49c6-9901-9cea5cc874ed)
> 
> Also after creating the realm, zonegroup and zone [screen-capture (11).webm](https://github.com/ceph/ceph/assets/106810838/97071f36-8562-4e0e-8378-c1844f511a6a)

Thanks @Pegonzal . I have addressed most of the comments

**Pull request closed without merging**